enclave/asym_keys.c

```diff
@@ -0,0 +1,553 @@
+        oe_free(key_buffer_local);
+    }
+
+    if (key_info_local != NULL)
+    {
+        oe_secure_zero_fill(key_info_local, key_info_size_local);
+        oe_free(key_info_local);
+    }
+
+    if (key != NULL)
+    {
+        oe_secure_zero_fill(key, key_size);
+        oe_free(key);
+    }
+
+    return result;
+}
+
+static oe_result_t _load_asymmetric_key(
+    const oe_asymmetric_key_params_t* key_params,
+    bool is_public,
+    const uint8_t* key_info,
+    size_t key_info_size,
+    uint8_t** key_buffer,
+    size_t* key_buffer_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    uint8_t* key = NULL;
+    size_t key_size = 0;
+    uint8_t* key_buffer_local = NULL;
+    size_t key_buffer_size_local = 0;
+
+    /* Check invalid params. */
+    if (!key_info || !key_buffer || !key_buffer_size)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    OE_CHECK(_check_asymmetric_key_params(key_params));
+
+    /* Load seal key. */
+    OE_CHECK(_load_seal_key(key_info, key_info_size, &key, &key_size));
+
+    /* Derive the asymmetric key. */
+    OE_CHECK(
+        _derive_asymmetric_key(
+            key_params,
+            is_public,
+            key,
+            key_size,
+            &key_buffer_local,
+            &key_buffer_size_local));
+
+    result = OE_OK;
+    *key_buffer = key_buffer_local;
+    *key_buffer_size = key_buffer_size_local;
+    key_buffer_local = NULL;
+
+done:
+    if (key_buffer_local != NULL)
+    {
+        oe_secure_zero_fill(key_buffer_local, key_buffer_size_local);
+        oe_free(key_buffer_local);
+    }
+
+    if (key != NULL)
+    {
+        oe_secure_zero_fill(key, key_size);
+        oe_free(key);
+    }
+
+    return result;
+}
+
+oe_result_t oe_get_public_key_by_policy(
+    oe_seal_policy_t policy,
+    const oe_asymmetric_key_params_t* key_params,
+    uint8_t** key_buffer,
+    size_t* key_buffer_size,
+    uint8_t** key_info,
+    size_t* key_info_size)
+{
+    return _load_asymmetric_key_by_policy(
+        policy,
+        key_params,
+        true,
+        key_buffer,
+        key_buffer_size,
+        key_info,
+        key_info_size);
+}
+
+oe_result_t oe_get_public_key(
+    const oe_asymmetric_key_params_t* key_params,
+    const uint8_t* key_info,
+    size_t key_info_size,
+    uint8_t** key_buffer,
+    size_t* key_buffer_size)
+{
+    return _load_asymmetric_key(
+        key_params, true, key_info, key_info_size, key_buffer, key_buffer_size);
+}
+
+oe_result_t oe_get_private_key_by_policy(
+    oe_seal_policy_t policy,
+    const oe_asymmetric_key_params_t* key_params,
+    uint8_t** key_buffer,
+    size_t* key_buffer_size,
+    uint8_t** key_info,
+    size_t* key_info_size)
+{
+    return _load_asymmetric_key_by_policy(
+        policy,
+        key_params,
+        false,
+        key_buffer,
+        key_buffer_size,
+        key_info,
+        key_info_size);
+}
+
+oe_result_t oe_get_private_key(
+    const oe_asymmetric_key_params_t* key_params,
```

Comment:
Is there anyway to get both keys in one call?

Reply:
the APIs we agreed upon had the private key/public key separation.

Reply:
For our use, we wouldn't use oe_get_keypair because what we actually want is the private key and the certificate that wraps the public key (not the raw public key itself).